### DISCUSSION TITLE
Builtin web search feature using various providers/engines

### DISCUSSION THREAD

**Discussion by secondtruth** [mckaywrigley/chatbot-ui#1359] [Ideas]
I'd like to propose an enhancement to include a builtin web search functionality using various search providers, globally available for every user. This feature would allow models to perform searches on behalf of the user, broadening the utility of the app. It could be enabled/disabled using a toggle switch near the prompt input box.

Suggestions and improvements of this proposal are always very appreciated!

### Suggested Search Providers
- [SearXNG](https://docs.searxng.org/dev/search_api.html)
- [YOU](https://documentation.you.com)
- [SerpAPI](https://serpapi.com)
- [Serper](https://serper.dev)
- [Serpstack](https://serpstack.com)

### Suggested Action Flow
1. Generating an appropriate search query from the user prompt.
2. Performing web search using the generated query and extracting content from found webpages.
3. Creating embeddings from website texts and perform Retrieval-Augmented Generation like with files.

## Related
- https://github.com/mckaywrigley/chatbot-ui/discussions/1159
- https://github.com/mckaywrigley/chatbot-ui/discussions/1191

**Comment by mysterium-coniunctionis** [comment]
I would love to see an integration with SearXNG.
I would love to see an integration with SearXNG.
I would love to see an integration with SearXNG.
I would love to see an integration with SearXNG.
I would love to see an integration with SearXNG.
I would love to see an integration with SearXNG.
I would love to see an integration with SearXNG.
I would love to see an integration with SearXNG.
I would love to see an integration with SearXNG.
**I would love to see an integration with SearXNG.**

_Any questions?_

(...especially because it is so easy to spin up a SearXNG instance on a home server and would eliminate or reduce worries about privacy and going over the SerpAPI free-tier limit for those of us using Chatbot-UI for personal and not business use.)

**Reply by secondtruth** [reply]
Hugging Face's chat-ui lately received a PR adding SearXNG integration. I also think search is a must have now, especially using SearXNG.
